cooked-validators/src/Cooked/MockChain/Monad.hs

```diff
@@ -157,6 +158,16 @@ spOutsFromCardanoTx cardanoTx = forM (Pl.getCardanoTxOutRefs cardanoTx) $
       Just chainIndexTxOut -> spOutResolveDatum (txOutRef, chainIndexTxOut)
       Nothing -> fail "could not extract ChainIndexTxOut"
 
+-- | Retrieve a typed datum from a 'SpendableOut'. This function is useful if you
+-- know the type of datum you expect on a 'SpendableOut' and you want to extract
+-- the datum directly to the correct 'DatumType' for your validator.
+--
+-- This function has an ambiguous type, so it is probably necessary to
+-- type-apply it to the type of your validator.
+spOutGetDatum :: Pl.FromData (Pl.DatumType a) => SpendableOut -> Maybe (Pl.DatumType a)
+spOutGetDatum (_, Pl.ScriptChainIndexTxOut _ _ (Right (Pl.Datum datum)) _) = Pl.fromBuiltinData datum
+spOutGetDatum _ = Nothing
```

Comment:
The `ScriptChainIndexTxOut` could contain a `Left datumhash` that is resolvable in a `MonadBlockChain`. We should not evaluate to `Nothing` in this case because it would imply there is no datum.

We could have precondition in the specs forbidding spOuts with datum hash but no explicit datum (e.g. after they have been processed with `spOutResolveDatum` in a `MonadBlockChain`). This would not be enforced at the type level though. But neither is the translation of `SpendsScript` constaint for the moment: they also require "well-formed-for-cooked" spouts.

Reply:
I was not suggesting to make it live in a blockchain (just stating it was resolvable in a blockchain) :smile: . I was suggesting to add a precondition in the specs, like the big warning in the `SpendsScript` constraint. I do not mind about your solution.

I think we do need to sort this all out though the whole code base though, for clarity and consistency. How to name/specify/type (e.g. have a `newtype` for spout with/without explicit datum)/make monadic or not/comment/etc.